### PR TITLE
Add support for georeferencing vector layers in georeferencer

### LINKED ISSUE #41300: [Georeferencer] Georeferencing vector data
**Feature description.**

When importing data without coordinate refenre system, such as CAD drawings, it is tedious to georeference the data, you must set the scale, then transform, then rorate, then transform again until everything is fine.

I forgot that arcthing had a tool for this https://desktop.arcgis.com/en/arcmap/10.3/manage-data/editing-existing-features/about-spatial-adjustment.htm

The georeferencer is currently based on gdalwarp which currently only supports raster data, but I think that by using the deformation in gdal it could be used to handle vector data. Though I'm not sure if this should be a feature request in gdal itself associated with this.

But I think that once we have a good alg to handle this it shouldn't be too hard to do the georeferencer tweaks in QGIS.

### PR DESCRIPTION
**Pull request by nyalldawson**
Allows vector layers without spatial referencing to be interactively georeferenced (or layers with referencing to be re-referenced!).

Georeferencing occurs in a task, so it remains nice and responsive even with massive layers. 

Sponsored by the Danish QGIS Usergroup

This PR builds on the work started by @roya0045 in https://github.com/qgis/QGIS/pull/41386 -- full credit to Alex for the groundwork here!

Fixes #41300

### CODE REVIEW

**Comment by nyalldawson** [comment]
@roya0045 would love some feedback if you have a chance to build and test this branch!

File: src/app/georeferencer/qgsgeorefmainwindow.cpp [hunk 1]
```diff
@@ -214,59 +216,126 @@ void QgsGeoreferencerMainWindow::openRaster( const QString &fileName )
   }
 
   QgsSettings s;
+  QString provider;
+  QString uri;
   if ( fileName.isEmpty() )
   {
-    QString dir = s.value( QStringLiteral( "/Plugin-GeoReferencer/rasterdirectory" ) ).toString();
+    QString dir = s.value( QStringLiteral( "/Plugin-GeoReferencer/lastdirectory" ) ).toString();
```
Comment (3nids) [inline src/app/georeferencer/qgsgeorefmainwindow.cpp:223]: can you create a setting in the registry for this?
I guess it makes sense to rename it at the same time.

File: src/app/georeferencer/qgsgeorefmainwindow.cpp [hunk 2]
```diff
@@ -214,59 +216,126 @@ void QgsGeoreferencerMainWindow::openRaster( const QString &fileName )
...
+    case QgsMapLayerType::AnnotationLayer:
+    case QgsMapLayerType::PointCloudLayer:
+    case QgsMapLayerType::GroupLayer:
+      break;
   }
 
-  QFileInfo fileInfo( mRasterFileName );
-  s.setValue( QStringLiteral( "/Plugin-GeoReferencer/rasterdirectory" ), fileInfo.path() );
+  QFileInfo fileInfo( mFileName );
+  s.setValue( QStringLiteral( "/Plugin-GeoReferencer/lastdirectory" ), fileInfo.path() );
```
Comment (3nids) [inline src/app/georeferencer/qgsgeorefmainwindow.cpp:305]: and reuse here

File: src/app/georeferencer/qgsgeorefmainwindow.cpp (hunk 1 above)
Comment (nyalldawson) [reply src/app/georeferencer/qgsgeorefmainwindow.cpp:223]: @3nids 
Can you give an example of how we should be doing that now?

File: src/app/georeferencer/qgsgeorefmainwindow.cpp [hunk 3]
```diff
@@ -1195,10 +1356,12 @@ void QgsGeoreferencerMainWindow::writeSettings()
...
 
   // warp options
-  s.setValue( QStringLiteral( "/Plugin-GeoReferencer/transformparam" ), static_cast< int >( mTransformParam ) );
-  s.setValue( QStringLiteral( "/Plugin-GeoReferencer/resamplingmethod" ), mResamplingMethod );
+  s.setValue( QStringLiteral( "/Plugin-GeoReferencer/transformmethod" ), qgsEnumValueToKey( mTransformMethod ) );
+  s.setValue( QStringLiteral( "/Plugin-GeoReferencer/resamplingmethod" ), qgsEnumValueToKey( mResamplingMethod ) );
   s.setValue( QStringLiteral( "/Plugin-GeoReferencer/compressionmethod" ), mCompressionMethod );
   s.setValue( QStringLiteral( "/Plugin-GeoReferencer/usezerofortrans" ), mUseZeroForTrans );
+  s.setValue( QStringLiteral( "/Plugin-GeoReferencer/saveGcpPoints" ), mSaveGcp );
+  s.setValue( QStringLiteral( "/Plugin-GeoReferencer/loadInProject" ), mLoadInQgis );
```
Comment (3nids) [inline src/app/georeferencer/qgsgeorefmainwindow.cpp:1364]: and for these

File: src/app/georeferencer/qgsimagewarper.h [hunk 4]
```diff
@@ -23,25 +23,40 @@
...
+    enum class Result
+    {
+      Success, //!< Warping completed successfully
+      Canceled, //!< Task was canceled before completion
+      InvalidParameters, //!< Invalid transform parameters
+      SourceError, //!< Error reading source
+      TransformError, //!< Error creating GDAL transformer
+      DestinationCreationError, //!< Error creating destination file
+      WarpFailure, //!< Failed warping source
     };
```
Comment (3nids) [inline src/app/georeferencer/qgsimagewarper.h:60]: make it Q_ENUM directly?

File: src/app/georeferencer/qgstransformsettingsdialog.cpp [hunk 5]
```diff
@@ -21,38 +21,63 @@
...
+    case QgsMapLayerType::AnnotationLayer:
+    case QgsMapLayerType::PointCloudLayer:
+    case QgsMapLayerType::GroupLayer:
+      break;
+  }
+  mOutputSettingsStackedWidget->adjustSize();
+  mOutputSettingsGroupBox->adjustSize();
+
+  outputFile->setDialogTitle( tr( "Destination File" ) );
+  outputFile->setDefaultRoot( settings.value( QStringLiteral( "/Plugin-GeoReferencer/lastSourceDir" ), QDir::homePath() ).toString() );
```
Comment (3nids) [inline src/app/georeferencer/qgstransformsettingsdialog.cpp:75]: another setting

**Comment by nyalldawson** [comment]
@3nids all done!

**Comment by roya0045** [comment]
> @roya0045 would love some feedback if you have a chance to build and test this branch!

I don't think I can provide timely feedback, the sooner I might be able to do is this weekend and that might be unrealistic. I don't doubt the quality of the changes, the only element that I had issues with my own version was the exporter. It converted any outputs to a geopackage and I couldn't figure out why this format was imposed despite calling the exporter 'properly'. 

Did you manage to find out what was causing this issue with the QgsVectorFileWriter?

**Comment by roya0045** [comment]
If that's fixed I assume the remainder is fine. 👍

File: src/app/georeferencer/qgsgeorefmainwindow.cpp [hunk 6]
```diff
@@ -1322,39 +1485,177 @@ bool QgsGeoreferencerMainWindow::georeference()
...
+        break;
+    }
+    return false;
+  }
+}
+
+bool QgsGeoreferencerMainWindow::georeferenceRaster()
+{
+
+  QgsImageWarperTask *task = new QgsImageWarperTask(
```
Comment (nirvn) [inline src/app/georeferencer/qgsgeorefmainwindow.cpp:1511]: I :heart: your move to georefence rasters through tasks.

File: src/app/georeferencer/qgsimagewarper.h [hunk 7]
```diff
@@ -23,26 +23,42 @@
...
+    Q_ENUM( Result )
 
     /**
      * Warp the file specified by "input" and write the resulting raster to the file "output".
```
Comment (nirvn) [inline src/app/georeferencer/qgsimagewarper.h:64]:
```suggestion
     * Warp the file specified by \a input and write the resulting raster to the file \a output.
```

There are a few params missing (pre-existing compression and crs as well as newly added feedback), maybe worth adding those.

File: src/analysis/georeferencing/qgsvectorwarper.h [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#ifndef QGSVECTORWARPER_H
+#define QGSVECTORWARPER_H
+
+#include <gdal_alg.h>
```
Comment (nirvn) [inline src/analysis/georeferencing/qgsvectorwarper.h:19]: Move below qgis source code include?

**Review by nirvn — APPROVED** [review]
@nyalldawson , I feel like an editor who's trying to justify his salary by pointing out frivolous issues here ;) small formatting comments added, couldn't find anything fundamentally wrong.

**Comment by nyalldawson** [comment]
Thanks @nirvn, all addressed!

**Comment by agiudiceandrea** [comment]
@nyalldawson, may this PR be the cause of https://github.com/qgis/QGIS/issues/51197 and https://github.com/qgis/QGIS/issues/51299?

**Comment by roya0045** [comment]
> @nyalldawson, may this PR be the cause of #51197 and #51299?

Would be more likely to have been caused by something during the rework ( the previous PR affecting the georeferencer I recall). Regardless, of when, there's something to fix.

**Comment by agiudiceandrea** [comment]
> Would be more likely to have been caused by something during the rework

If you are referring to https://github.com/qgis/QGIS/pull/47141 ("Fix lots of georeferencer issues") and https://github.com/qgis/QGIS/pull/47279 ("Georeferencer fixes, pt 2"), they have been both allegedly backported to 3.22 (https://github.com/qgis/QGIS/pull/47276), so the issue should have occurred also in version 3.22, while it seems not.

**Pull request merged**